std/algebra/emulated/sw_bw6761/pairing.go

```diff
@@ -255,114 +261,50 @@ func (pr Pairing) MillerLoop(P []*G1Affine, Q []*G2Affine) (*GTEl, error) {
-					R1: *pr.curveF.MulMod(&l1.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l1.R1, &l1.R0)
-			case -1:
-				accQ[k], l0, l1 = pr.doubleAndAddStep(accQ[k], negQ[k])
-				l0 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l0.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l0.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l0.R1, &l0.R0)
-				l1 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l1.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l1.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l1.R1, &l1.R0)
-			case 0:
-				accQ[k], l0 = pr.doubleStep(accQ[k])
-				l0 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l0.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l0.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l0.R1, &l0.R0)
-			case 1:
-				accQ[k], l0, l1 = pr.doubleAndAddStep(accQ[k], Q[k])
-				l0 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l0.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l0.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l0.R1, &l0.R0)
-				l1 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l1.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l1.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l1.R1, &l1.R0)
-			case 3:
-				accQ[k], l0, l1 = pr.doubleAndAddStep(accQ[k], imQ[k])
-				l0 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l0.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l0.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l0.R1, &l0.R0)
-				l1 = &lineEvaluation{
-					R0: *pr.curveF.MulMod(&l1.R0, xNegOverY[k]),
-					R1: *pr.curveF.MulMod(&l1.R1, yInv[k]),
-				}
-				result = pr.MulBy014(result, &l1.R1, &l1.R0)
-			default:
-				return nil, errors.New("invalid loopCounter")
+			result = pr.MulBy014(result,
+				pr.curveF.MulMod(&lines[k][0][i].R1, yInv[k]),
+				pr.curveF.MulMod(&lines[k][0][i].R0, xNegOverY[k]),
+			)
+		}
+
+		if loopCounter2[i]*3+loopCounter1[i] != 0 {
+			for k := 0; k < n; k++ {
+				result = pr.MulBy014(result,
+					pr.curveF.MulMod(&lines[k][1][i].R1, yInv[k]),
+					pr.curveF.MulMod(&lines[k][1][i].R0, xNegOverY[k]),
+				)
 			}
 		}
 	}
 
-	// i = 0, j = -3
-	// The resulting accumulator point is the infinity point because
-	// [(x₀+1) + λ(x₀³-x₀²-x₀)]Q = [3(x₀-1)² ⋅ r]Q = ∞
-	// since we're using affine coordinates, the addStep in the last iteration
-	// (j=-3) will fail as the slope of a vertical line in indefinite. But in
-	// projective coordinates, vertinal lines meet at (0:1:0) so the result
-	// should be unchanged if we ommit the addStep in this case. Moreover we
-	// just compute before the tangent line and not the full doubleStep as we
-	// only care about the Miller loop result in Fp6 and not the point itself.
 	result = pr.Square(result)
 	for k := 0; k < n; k++ {
-		l0 = pr.tangentCompute(accQ[k])
-		l0 = &lineEvaluation{
-			R0: *pr.curveF.MulMod(&l0.R0, xNegOverY[k]),
-			R1: *pr.curveF.MulMod(&l0.R1, yInv[k]),
-		}
-		result = pr.MulBy014(result, &l0.R1, &l0.R0)
+		result = pr.MulBy014(result,
```

Comment:
I think we can keep this iteration inside the loop because the special case is handled by `precomputeLines` (with `tangentCompute`)

Reply:
Merged the iterations. But as I'm not computing second line at all for `i=0`, then added additional conditional `i>0` here.